File: probs/279.cpp
class Solution {
public:
    int numSquares(int n) {
        int memo[10001];
        for (int idx = 0; idx <= 10000; idx++) {
            memo[idx] = idx;
        }

        for (int num = 2; num * num <= 10000; num++) {
            int sp = num * num;
            memo[sp] = 1;

            for (int idx = sp + 1; idx <= 10000; idx++) {
                if (memo[idx] > memo[idx - sp] + memo[sp]) {
                    memo[idx] = memo[idx - sp] + memo[sp];
                }
            }
        }

        return memo[n];
    }
};

File: probs/279.rs
impl Solution {
    pub fn num_squares(n: i32) -> i32 {
        let mut memo: [i32; 10001] = [0; 10001];
        
        for idx in 0..10001 {
            memo[idx] = idx as i32;
        }

        let mut num = 2;

        while num * num <= 10000 {
            let mut start = num * num;
            memo[start] = 1;

            while start <= 10000 {
                if memo[start] > memo[start - num * num] + 1 {
                    memo[start] = memo[start - num * num] + 1;
                }

                start += 1;
            }

            num += 1;
        }

        return memo[n as usize];
    }
}